packages/statemanager/src/accessWitness.ts

```diff
@@ -0,0 +1,433 @@
+    gas += this.touchAddressOnWriteAndComputeGas(origin, 0, BALANCE_LEAF_KEY)
+
+    return gas
+  }
+
+  touchTxExistingAndComputeGas(target: Address, { sendsValue }: { sendsValue?: boolean } = {}) {
+    let gas = BIGINT_0
+
+    gas += this.touchAddressOnReadAndComputeGas(target, 0, VERSION_LEAF_KEY)
+    gas += this.touchAddressOnReadAndComputeGas(target, 0, CODE_SIZE_LEAF_KEY)
+    gas += this.touchAddressOnReadAndComputeGas(target, 0, CODE_KECCAK_LEAF_KEY)
+    gas += this.touchAddressOnReadAndComputeGas(target, 0, NONCE_LEAF_KEY)
+
+    if (sendsValue === true) {
+      gas += this.touchAddressOnWriteAndComputeGas(target, 0, BALANCE_LEAF_KEY)
+    } else {
+      gas += this.touchAddressOnReadAndComputeGas(target, 0, BALANCE_LEAF_KEY)
+    }
+
+    return gas
+  }
+
+  touchCodeChunksRangeOnReadAndChargeGas(contact: Address, startPc: number, endPc: number) {
+    let gas = BIGINT_0
+    for (let chunkNum = Math.floor(startPc / 31); chunkNum <= Math.floor(endPc / 31); chunkNum++) {
+      const { treeIndex, subIndex } = getTreeIndicesForCodeChunk(chunkNum)
+      gas += this.touchAddressOnReadAndComputeGas(contact, treeIndex, subIndex)
+    }
+    return gas
+  }
+
+  touchCodeChunksRangeOnWriteAndChargeGas(contact: Address, startPc: number, endPc: number) {
+    let gas = BIGINT_0
+    for (let chunkNum = Math.floor(startPc / 31); chunkNum <= Math.floor(endPc / 31); chunkNum++) {
+      const { treeIndex, subIndex } = getTreeIndicesForCodeChunk(chunkNum)
+      gas += this.touchAddressOnWriteAndComputeGas(contact, treeIndex, subIndex)
+    }
+    return gas
+  }
+
+  touchAddressOnWriteAndComputeGas(
+    address: Address,
+    treeIndex: number | bigint,
+    subIndex: number | Uint8Array
+  ): bigint {
+    return this.touchAddressAndChargeGas(address, treeIndex, subIndex, { isWrite: true })
+  }
+
+  touchAddressOnReadAndComputeGas(
+    address: Address,
+    treeIndex: number | bigint,
+    subIndex: number | Uint8Array
+  ): bigint {
+    return this.touchAddressAndChargeGas(address, treeIndex, subIndex, { isWrite: false })
+  }
+
+  touchAddressAndChargeGas(
+    address: Address,
+    treeIndex: number | bigint,
+    subIndex: number | Uint8Array,
+    { isWrite }: { isWrite?: boolean }
+  ): bigint {
+    let gas = BIGINT_0
+
+    const { stemRead, stemWrite, chunkRead, chunkWrite, chunkFill } = this.touchAddress(
+      address,
+      treeIndex,
+      subIndex,
+      { isWrite }
+    )
+
+    if (stemRead) {
+      gas += WitnessBranchReadCost
+    }
+    if (stemWrite) {
+      gas += WitnessBranchWriteCost
+    }
+
+    if (chunkRead) {
+      gas += WitnessChunkReadCost
+    }
+    if (chunkWrite) {
+      gas += WitnessChunkWriteCost
+    }
+    if (chunkFill) {
+      gas += WitnessChunkFillCost
+    }
+
+    debug(
+      `touchAddressAndChargeGas=${gas} address=${address} treeIndex=${treeIndex} subIndex=${subIndex}`
+    )
+
+    return gas
+  }
+
+  touchAddress(
```

Comment:
Just to note:

there is an insane amount of new "Codecov - not covered by tests" messages in this PR.

Know about the experimental state of all this, maybe it would nevertheless make sense to put some emphasis on this relatively soon, to have some internal unit tests to get all this better covered (potentially also a good learning task for others from the team to contribute).

Reply:
I will take this on to get a bit more deeper in this code! :+1: